types/externs.d.ts

```diff
@@ -51,30 +27,13 @@ declare global {
   };
 
   /** Make optional all properties on T and any properties on object properties of T. */
-  type RecursivePartial<T> = {
-    [P in keyof T]+?:
-      // RE: First two conditions.
-      // If type is a union, map each individual component and transform the resultant tuple back into a union.
-      // Only up to 4 components of a union is supported (all but the last few are dropped). For more, modify the second condition
-      // and `IntersectionOfFunctionsToType`.
-      // Ex: `{passes: PassJson[] | null}` - T[P] doesn't exactly match the array-recursing condition, so without these first couple
-      // conditions, it would fall through to the last condition (would just return T[P]).
-
-      // RE: First condition.
-      // Guard against large string unions, which would be unreasonable to support (much more than 4 components is common).
-
-      SplitType<T[P]> extends string[] ? T[P] :
-      GetLength<SplitType<T[P]>> extends 2|3|4 ? RecursivePartialUnion<T[P]>[number] :
-
-      // Recurse into arrays.
-      T[P] extends (infer U)[] ? RecursivePartial<U>[] :
-
-      // Recurse into objects.
-      T[P] extends (object|undefined) ? RecursivePartial<T[P]> :
-
-      // Strings, numbers, etc. (terminal types) end here.
-      T[P];
-  };
+  type RecursivePartial<T> =
```

Comment:
Hey, now *here's* a type I can understand 😆 

I'm curious how this solves the original `type Foo = {foo: Array<T> | null}` problem though...

> As long as we keep the conditional type distributing across unions correctly, 

Where does that happen? Does typescript do this for us magically somehow? AFAICT this just applies to arrays and straight up objects and everything else just returns `T`, so wouldn't `RecursivePartial<Foo>` just be `{foo?: Array<T> | null}` instead of `{foo?: Array<RecursivePartial<T>> | null}`?

> they don't need to be split up into tuples and merged back again.

Hallelujah 🎉

Reply:
> Does typescript do this for us magically somehow?

Basically, yes, but in a mostly good way :) I should probably have explained why that works now, though.

They [define it](https://www.typescriptlang.org/docs/handbook/advanced-types.html#distributive-conditional-types) kind of poorly, but as long as the type parameter is used directly as part of the check, the conditional type will distribute across unions.

> `type Foo = {foo: Array<T> | null}`
> `RecursivePartial<Foo>`

So (if working correctly):
- `Foo` goes into the `object` branch, `Foo.foo` is made optional by the `?:`, then `Foo.foo`'s type is passed into `RecursivePartial`.
- `RecursivePartial<Array<T> | null>` would distribute and be treated as `RecursivePartial<Array<T>> | RecursivePartial<null>`, giving `Array<RecursivePartial<T>> | null`.

In this case the previous code was checking the conditionals against `T[P]`, which was preventing distribution. By checking directly against `T` instead (so all the `T[P]` checks move one layer deeper and are themselves properly distributed since they're also looked at individually), it's distributive again.

Reply:
The other type of conditional that breaks distribution is if the type parameter isn't the thing being checked against.

The `RecursivePartialArrayOrTuple` I suggested below would fail both checks because it's doing `number extends T['length'] ?`. `T` isn't being used directly and it's `number` that's been checked, not the type parameter (to check for bounded vs unbounded lengths).

So if we do end up using something like that, it should only be called by `RecursivePartial`, which would have already done the distributing and can call `RecursivePartialArrayOrTuple` on what it knows are non-union types.

Reply:
gotcha, yeah that makes sense 👍  

I guess I never really questioned `Partial<ObjA | ObjB>` distribution either but that's probably only because I never saw a 50 line implementation of it that I thought was necessary :)